22.0.0.12-beta staging (#2681)
OpenLiberty/blogs: posts/2022-10-27-22.0.0.12-beta.adoc

blog_description: "Open Liberty continues to make progress towards full Jakarta EE 10 compliance with the debut of six new specification implementations Jakarta WebSocket 2.1, Jakarta Batch 2.1, Jakarta XML Web Services 4.0, Jakarta Messaging 3.1, Jakarta Messaging 3.1, Jakarta Pages 3.1 and Jakarta Tags 3.0.  Two new MicroProfile 6 features are also introduced, MicroProfile JSON Web Token 2.1 and MicroProfile Metrics 5.0."
open-graph-image: https://openliberty.io/img/twitter_card.jpg
open-graph-image-alt: Open Liberty Logo
---
= More Jakarta EE 10 and MicroProfile 6 features make their debut in Open Liberty 22.0.0.12-beta!
Michal Broz <https://github.com/mbroz2>
:imagesdir: /
:url-prefix:
:url-about: /
//Blank line here is necessary before starting the body of the post.

> As we continue to make progress towards full Jakarta EE 10 compliance, the 22.0.0.12-beta release brings with it six new Jakarta EE 10 specifications, alongside the ones which have been part of previous beta releases.  Jakarta Batch 2.1 provides improved CDI integration while Jakarta XML Web Services 4.0 now incorporates Jakarta Web Services Metadata.  Jakarta Pages 3.1 allows developers to identify undefined Expression Language variables and raise an Exception, while Jakarta Tags 3.0 includes Tag URI renaming when referencing tag libraries. Jakarta Messaging 3.1 makes connection factory and destination definition annotations repeatable, while Jakarta WebSocket 2.1 allows for the registration of endpoints outside of the deployment phase.

Reviewer: ```suggestion
As we progress towards full Jakarta EE 10 compliance, the 22.0.0.12-beta release includes six new Jakarta EE 10 specifications, alongside the ones that were included in previous beta releases.  

Jakarta Batch 2.1 provides improved CDI integration while Jakarta XML Web Services 4.0 now incorporates Jakarta Web Services Metadata.  Jakarta Pages 3.1 allows developers to identify undefined Expression Language variables and raise an Exception, while Jakarta Tags 3.0 includes Tag URI renaming when referencing tag libraries. Jakarta Messaging 3.1 makes connection factory and destination definition annotations repeatable, while Jakarta WebSocket 2.1 allows for the registration of endpoints outside of the deployment phase.

```

> As we continue to make progress towards full Jakarta EE 10 compliance, the 22.0.0.12-beta release brings with it six new Jakarta EE 10 specifications, alongside the ones which have been part of previous beta releases.  Jakarta Batch 2.1 provides improved CDI integration while Jakarta XML Web Services 4.0 now incorporates Jakarta Web Services Metadata.  Jakarta Pages 3.1 allows developers to identify undefined Expression Language variables and raise an Exception, while Jakarta Tags 3.0 includes Tag URI renaming when referencing tag libraries. Jakarta Messaging 3.1 makes connection factory and destination definition annotations repeatable, while Jakarta WebSocket 2.1 allows for the registration of endpoints outside of the deployment phase.

Reviewer 2: ```suggestion
As we continue to make progress towards full Jakarta EE 10 compliance, the 22.0.0.12-beta release brings with it six new Jakarta EE 10 specifications, alongside the ones which have been part of previous beta releases.  Jakarta Batch 2.1 provides improved CDI integration while Jakarta XML Web Services 4.0 now incorporates Jakarta Web Services Metadata.  Jakarta Pages 3.1 provides developers an option to raise Exceptions for undefined Expression Language Identifiers, while Jakarta Tags 3.0 includes Tag URI renaming when referencing tag libraries. Jakarta Messaging 3.1 makes connection factory and destination definition annotations repeatable, while Jakarta WebSocket 2.1 allows for the registration of endpoints outside of the deployment phase.

```

> But it's not only Jakarta EE 10 features that have made this beta, as two new MicroProfile 6 features are also included! MicroProfile JSON Web Token 2.1 brings with it simplified JWT validation and MicroProfile Metrics 5.0 provides user defined metric registry scopes.

Reviewer: ```suggestion
But this beta includes more than just Jakarta EE 10 features; two new MicroProfile 6.0 features are also included! MicroProfile JSON Web Token 2.1 includes simplified JWT validation and MicroProfile Metrics 5.0 provides user-defined metric registry scopes.

```

The link:{url-about}[Open Liberty] 22.0.0.12-beta includes the following beta features (along with link:{url-prefix}/docs/latest/reference/feature/feature-overview.html[all GA features]):

* Jakarta EE 10:

> ** <<pages, Identify undefined Expression Language variables and raise Exception (Jakarta Pages 3.1 and Jakarta Tags 3.0)>>

Reviewer 2: ```suggestion
** <<pages, Option to raise Exceptions for undefined Expression Language Identifiers (Jakarta Pages 3.1 and Jakarta Tags 3.0)>>

```

** <<websocket, Register endpoints outside of the deployment phase (Jakarta WebSocket 2.1)>>
** <<batch, Improved CDI integration (Jakarta Batch 2.1)>>
** <<messaging, Connection factory and destination definition annotations are now repeatable (Jakarta Messaging 3.1)>>
** <<xmlws, Jakarta Web Services Metadata (Jakarta XML Web Services 4.0)>>
* MicroProfile 6:
** <<jwt, Simplified JWT Validation (MicroProfile JSON Web Token 2.1)>>
** <<metrics, User defined metric registry scopes (MicroProfile Metrics 5.0)>>

See also link:{url-prefix}/blog/?search=beta&key=tag[previous Open Liberty beta blog posts].

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // // 
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/22563
// Contact/Reviewer: volosied,pnicolucci,ReeceNana
// // // // // // // // 
[#pages]

> == Identify undefined Expression Language variables and raise Exception (Jakarta Pages 3.1 and Jakarta Tags 3.0)

Reviewer 3: Is this it title document of just a highlight of new feature? 

Looking over it again, I would like to reword it.  Something like: `Option to Raise Exceptions for undefined Expression Language Identifiers"

> == Identify undefined Expression Language variables and raise Exception (Jakarta Pages 3.1 and Jakarta Tags 3.0)

Reviewer 2: ```suggestion
== Option to raise Exceptions for undefined Expression Language Identifiers (Jakarta Pages 3.1 and Jakarta Tags 3.0)

```

> Jakarta Pages is a template engine that allows developers to mix various content, such as HTML, XML, and java code, and have it translated into a Jakarta Servlet. Jakarta Standard Tag Library allows for tags to encapsulate common functionality used in many web applications. The Open Liberty `pages-3.1` feature is an implementation of both the Jakarta Pages 3.1 and Jakarta Standard Tag Library 3.0 specifications. It includes new features, behavior clarifications, and deprecations from the previous 3.0 version.

Reviewer: ```suggestion
Jakarta Pages is a template engine that allows developers to mix various content, such as HTML, XML, and Java code, and translate it into a Jakarta Servlet. Jakarta Standard Tag Library allows for tags to encapsulate common functionality used in many web applications. The Open Liberty `pages-3.1` feature is an implementation of both the Jakarta Pages 3.1 and Jakarta Standard Tag Library 3.0 specifications. It includes new features, behavior clarifications, and deprecations from the previous 3.0 version.

```

> The new `pages-3.1` feature allows developers to identify undefined Expression Language variables via a new option, `errorOnELNotFound`/`<error-on-el-not-found>`, which raises a `PropertyNotFoundException`. It can be set via a page/tag directive or the web.xml, respectfully.

Reviewer: ```suggestion
The new `pages-3.1` feature allows developers to identify undefined Expression Language variables by using a new option, which raises a `PropertyNotFoundException` and can be set in one of two ways. The `errorOnELNotFound` attribute is set as a page or tag directive.  The  `<error-on-el-not-found>` element is set in  the web.xml file.

```

[source,xml]
---- 
<featureManager>
   <feature>pages-3.1</feature>
</featureManager>
---- 
 
For more information about Jakarta Server Pages and Jakarta Standard Tag Library, refer to the following links:  

* link:{url-prefix}/docs/latest/reference/feature/pages.html[Feature Documentation: Jakarta Server Pages]
* link:https://jakarta.ee/specifications/pages/3.1/[Specification: Jakarta Server Pages 3.1]
* link:https://jakarta.ee/specifications/tags/3.0/[Specification: Jakarta Standard Tag Library 3.0]
* link:https://jakarta.ee/specifications/pages/3.1/apidocs/jakarta.servlet.jsp/module-summary.html[Javadoc: Jakarta Server Pages 3.1]
* link:https://jakarta.ee/specifications/tags/3.0/apidocs/jakarta.servlet.jsp.jstl/module-summary.html[Javadoc: Jakarta Standard Tag Library 3.0]

// DO NOT MODIFY THIS LINE. </GHA-BLOG-TOPIC> 

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // // 
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/22416
// Contact/Reviewer: volosied,pnicolucci,ReeceNana
// // // // // // // // 
[#websocket]
== Register endpoints outside of the deployment phase (Jakarta WebSocket 2.1)

> Jakarta WebSocket allows communication via the WebSocket protocol, allowing for full-duplex communication between endpoints.  This means that, unlike the HTTP protocol, a single, persisted connection is made across the endpoints, that allows the server to continuously push information to the client. 

Reviewer 3: Don't like this wording after another read through. Can we change it to: 

This means that a single, persisted connection is made across the endpoints, allowing both the server and client to continuously and simultaneously push information to each other.

Reviewer 2: I was trying to focus on the differentiation from HTTP, but this makes sense

> Jakarta WebSocket allows communication via the WebSocket protocol, allowing for full-duplex communication between endpoints.  This means that, unlike the HTTP protocol, a single, persisted connection is made across the endpoints, that allows the server to continuously push information to the client. 

Reviewer: ```suggestion
Jakarta WebSocket allows full-duplex communication between endpoints via the WebSocket protocol. This means that, unlike the HTTP protocol, a single, persisted connection is made across the endpoints, which allows the server to continuously push information to the client. 
```

> Jakarta WebSocket allows communication via the WebSocket protocol, allowing for full-duplex communication between endpoints.  This means that, unlike the HTTP protocol, a single, persisted connection is made across the endpoints, that allows the server to continuously push information to the client. 

Reviewer 2: ```suggestion
Jakarta WebSocket allows communication via the WebSocket protocol, allowing for full-duplex communication between endpoints.  This means that a single, persisted connection is made across the endpoints, allowing both the server and client to continuously and simultaneously push information to each other.

```

> Jakarta WebSocket 2.1 is the latest version of the specification and provides several new features and behavior clarifications aimed at improving websocket development. Firstly, when setting up dependencies, developers are likely to notice the API jars have been refactored into client and server jars; however, both APIs are available as part of the `websocket-2.1` feature.

Reviewer: ```suggestion
Jakarta WebSocket 2.1 is the latest version of the specification and provides several new features and behavior clarifications aimed at improving WebSocket development. Firstly, when setting up dependencies, developers are likely to notice the API JARs are refactored into client and server JARs. However, both APIs are available as part of the `websocket-2.1` feature.

```

> Developers can now register endpoints outside of the deployment phase of the web application, where previously there was a restriction preventing this. HTTP connection upgrades to the websocket protocol can now occur via the link:https://jakarta.ee/specifications/websocket/2.1/apidocs/server/jakarta/websocket/server/servercontainer#upgradeHttpToWebSocket(java.lang.Object,java.lang.Object,jakarta.websocket.server.ServerEndpointConfig,java.util.Map)[`ServerContainer.upgradeHttpToWebSocket()` method​] rather than IBM's specific link:https://www.ibm.com/docs/api/v1/content/SSD28V_liberty/com.ibm.websphere.javadoc.liberty.doc/com.ibm.websphere.appserver.api.wsoc_1.0-javadoc/com/ibm/websphere/wsoc/WsWsocServerContainer.html#doUpgrade(javax.servlet.http.HttpServletRequest,javax.servlet.http.HttpServletResponse,javax.websocket.server.ServerEndpointConfig,java.util.Map)[`WsWsocServerContainer.doUpgrade()` method].

Reviewer 3: Could you remove the first sentence: `Developers can now register endpoints outside of the deployment phase of the web application, where previously there was a restriction preventing this`  This was erroneously included in the beta blog but wasn't addressed in the websocket 2.1 code.  This feature will be included in the next release.

> Developers can now register endpoints outside of the deployment phase of the web application, where previously there was a restriction preventing this. HTTP connection upgrades to the websocket protocol can now occur via the link:https://jakarta.ee/specifications/websocket/2.1/apidocs/server/jakarta/websocket/server/servercontainer#upgradeHttpToWebSocket(java.lang.Object,java.lang.Object,jakarta.websocket.server.ServerEndpointConfig,java.util.Map)[`ServerContainer.upgradeHttpToWebSocket()` method​] rather than IBM's specific link:https://www.ibm.com/docs/api/v1/content/SSD28V_liberty/com.ibm.websphere.javadoc.liberty.doc/com.ibm.websphere.appserver.api.wsoc_1.0-javadoc/com/ibm/websphere/wsoc/WsWsocServerContainer.html#doUpgrade(javax.servlet.http.HttpServletRequest,javax.servlet.http.HttpServletResponse,javax.websocket.server.ServerEndpointConfig,java.util.Map)[`WsWsocServerContainer.doUpgrade()` method].

Reviewer: ```suggestion
Developers can now register endpoints outside of the deployment phase of the web application, which was previously not possible. HTTP connection upgrades to the WebSocket protocol can now occur via the link:https://jakarta.ee/specifications/websocket/2.1/apidocs/server/jakarta/websocket/server/servercontainer#upgradeHttpToWebSocket(java.lang.Object,java.lang.Object,jakarta.websocket.server.ServerEndpointConfig,java.util.Map)[ServerContainer.upgradeHttpToWebSocket() method​], rather than the IBM link:https://www.ibm.com/docs/api/v1/content/SSD28V_liberty/com.ibm.websphere.javadoc.liberty.doc/com.ibm.websphere.appserver.api.wsoc_1.0-javadoc/com/ibm/websphere/wsoc/WsWsocServerContainer.html#doUpgrade(javax.servlet.http.HttpServletRequest,javax.servlet.http.HttpServletResponse,javax.websocket.server.ServerEndpointConfig,java.util.Map)[WsWsocServerContainer.doUpgrade() method].

```

> Developers can now register endpoints outside of the deployment phase of the web application, where previously there was a restriction preventing this. HTTP connection upgrades to the websocket protocol can now occur via the link:https://jakarta.ee/specifications/websocket/2.1/apidocs/server/jakarta/websocket/server/servercontainer#upgradeHttpToWebSocket(java.lang.Object,java.lang.Object,jakarta.websocket.server.ServerEndpointConfig,java.util.Map)[`ServerContainer.upgradeHttpToWebSocket()` method​] rather than IBM's specific link:https://www.ibm.com/docs/api/v1/content/SSD28V_liberty/com.ibm.websphere.javadoc.liberty.doc/com.ibm.websphere.appserver.api.wsoc_1.0-javadoc/com/ibm/websphere/wsoc/WsWsocServerContainer.html#doUpgrade(javax.servlet.http.HttpServletRequest,javax.servlet.http.HttpServletResponse,javax.websocket.server.ServerEndpointConfig,java.util.Map)[`WsWsocServerContainer.doUpgrade()` method].

Reviewer 2: ```suggestion
HTTP connection upgrades to the WebSocket protocol can now occur via the link:https://jakarta.ee/specifications/websocket/2.1/apidocs/server/jakarta/websocket/server/servercontainer#upgradeHttpToWebSocket(java.lang.Object,java.lang.Object,jakarta.websocket.server.ServerEndpointConfig,java.util.Map)[`ServerContainer.upgradeHttpToWebSocket()` method​] rather than IBM's specific link:https://www.ibm.com/docs/api/v1/content/SSD28V_liberty/com.ibm.websphere.javadoc.liberty.doc/com.ibm.websphere.appserver.api.wsoc_1.0-javadoc/com/ibm/websphere/wsoc/WsWsocServerContainer.html#doUpgrade(javax.servlet.http.HttpServletRequest,javax.servlet.http.HttpServletResponse,javax.websocket.server.ServerEndpointConfig,java.util.Map)[`WsWsocServerContainer.doUpgrade()` method].

```

> Another change regards `Session#getUserProperties()`, as the returned map is now a copy per endpoint instance (per websocket session). Additionally, the server session's user properties are initially populated by `ServerEndpointConfig.getUserProperties()`, thus allowing endpoints to retrieve properties set in link:https://jakarta.ee/specifications/websocket/2.1/apidocs/server/jakarta/websocket/server/serverendpointconfig.configurator#modifyHandshake(jakarta.websocket.server.ServerEndpointConfig,jakarta.websocket.server.HandshakeRequest,jakarta.websocket.HandshakeResponse)[ServerEndpointConfig.Configurator.modifyHandshake() method​]. This is similar to the client session as the user properties are a shallow copy from link:https://jakarta.ee/specifications/websocket/2.1/apidocs/server/jakarta/websocket/EndpointConfig.html#getUserProperties()[`EndpointConfig.getUserProperties()`].

Reviewer: ```suggestion
Another change regards the `Session#getUserProperties()` method. The returned map is now a copy per endpoint instance, per WebSocket session. Additionally, the server session's user properties are initially populated by the `ServerEndpointConfig.getUserProperties()` method, so endpoints can retrieve properties set in the link:https://jakarta.ee/specifications/websocket/2.1/apidocs/server/jakarta/websocket/server/serverendpointconfig.configurator#modifyHandshake(jakarta.websocket.server.ServerEndpointConfig,jakarta.websocket.server.HandshakeRequest,jakarta.websocket.HandshakeResponse)[ServerEndpointConfig.Configurator.modifyHandshake() method​]. This is similar to the client session as the user properties are a shallow copy from the link:https://jakarta.ee/specifications/websocket/2.1/apidocs/server/jakarta/websocket/EndpointConfig.html#getUserProperties()[`EndpointConfig.getUserProperties()`] method.
```

> Lastly, there were many clarifications in the spec, such as requiring [Session.getRequestURI()](https://jakarta.ee/specifications/websocket/2.1/apidocs/server/jakarta/websocket/session#getRequestURI()) to return the whole URI, and these clarifications can be viewed on the [Jakarta WebSocket 2.1](https://jakarta.ee/specifications/websocket/2.1/) page.

Reviewer: ```suggestion
Lastly, there were many clarifications in the spec, such as requiring the link:https://jakarta.ee/specifications/websocket/2.1/apidocs/server/jakarta/websocket/session#getRequestURI()[Session.getRequestURI()] method to return the whole URI. These clarifications can be viewed on the link:https://jakarta.ee/specifications/websocket/2.1/[Jakarta WebSocket 2.1] page.

```

* link:https://jakarta.ee/specifications/websocket/2.1/apidocs/client[Client Javadoc: Jakarta WebSocket]
* link:https://jakarta.ee/specifications/websocket/2.1/apidocs/server[Server Javadoc: Jakarta WebSocket]

// DO NOT MODIFY THIS LINE. </GHA-BLOG-TOPIC> 

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // // 
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/22950
// Contact/Reviewer: ReeceNana,cgianfra
// // // // // // // // 
[#batch]
== Improved CDI integration (Jakarta Batch 2.1)
The updates in Jakarta Batch 2.1 primarily focus on improved integration with Context Dependency Injection (CDI). Discovery of Batch artifacts is now standardized via CDI, whereas before it was not fully defined in the specification.

Previously, the default `JobOperator` was only made available from a static factory method, `BatchRuntime.getJobOperator()`. With Jakarta Batch 2.1, the default `JobOperator` is provided as a CDI bean unless one is defined by the user application, if an injectable field is available as defined by: 
[source,java]
----
`@Inject` 
`JobOperator jobOperator;`
----

> Finally, the `@BatchProperty` annotation allowed for injection of String-type values only, but has been expanded to most other primitive types. For example:

Reviewer: ```suggestion
Previously, the `@BatchProperty` annotation allowed for injection of String-type values only, but is now expanded to most other primitive types. For example:

```

   <feature>batchManagement-1.0</feature>
   <feature>servlet-6.0</feature>
</featureManager>
----

For more information regarding Jakarta Batch 2.1, refer to the following links:  

* link:{url-prefix}/docs/latest/reference/feature/batch.html[Feature Documentation: Jakarta Batch]
* link:https://jakarta.ee/specifications/batch/2.1/jakarta-batch-spec-2.1.html[Specification: Jakarta Batch 2.1]

// DO NOT MODIFY THIS LINE. </GHA-BLOG-TOPIC> 

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // // 
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/22632
// Contact/Reviewer: AndrewDJBanks,ReeceNana
// // // // // // // // 
[#messaging]
== Connection factory and destination definition annotations are now repeatable (Jakarta Messaging 3.1)
Jakarta Messaging is a set of APIs to standardize message communication between applications, defining how the messages are created, sent, received, and interpreted.  Jakarta Messaging 3.1 provides minor changes and enhancements to the specification.  Two noteworthy updates are the requirement for Java 11 and making the `@JMSConnectionFactoryDefinition` and `@JMSDestinationDefinition` annotations repeatable.

> To enable the new Jakarta Messaging 4.0 feature, add the `messaging-3.1` feature to your server.xml:

Reviewer: ```suggestion
To enable the new Jakarta Messaging 3.1 feature, add the `messaging-3.1` feature to your server.xml:

```

But it's not only Jakarta EE 10 features that have made this beta, as two new MicroProfile 6 features are also included! MicroProfile JSON Web Token 2.1 brings with it simplified JWT validation and MicroProfile Metrics 5.0 provides user defined metric registry scopes.

The link:{url-about}[Open Liberty] 22.0.0.12-beta includes the following beta features (along with link:{url-prefix}/docs/latest/reference/feature/feature-overview.html[all GA features]):

* Jakarta EE 10:
** <<pages, Option to raise Exceptions for undefined Expression Language Identifiers (Jakarta Pages 3.1 and Jakarta Tags 3.0)>>

** <<websocket, Register endpoints outside of the deployment phase (Jakarta WebSocket 2.1)>>
** <<batch, Improved CDI integration (Jakarta Batch 2.1)>>
** <<messaging, Connection factory and destination definition annotations are now repeatable (Jakarta Messaging 3.1)>>
** <<xmlws, Jakarta Web Services Metadata (Jakarta XML Web Services 4.0)>>
* MicroProfile 6:
** <<jwt, Simplified JWT Validation (MicroProfile JSON Web Token 2.1)>>

> ** <<metrics, User defined metric registry scopes (MicroProfile Metrics 5.0)>>

Reviewer 2: ```suggestion
** <<metrics, User-defined metric registry scopes (MicroProfile Metrics 5.0)>>

```

* link:https://jakarta.ee/specifications/websocket/2.1/apidocs/client[Client Javadoc: Jakarta WebSocket]
* link:https://jakarta.ee/specifications/websocket/2.1/apidocs/server[Server Javadoc: Jakarta WebSocket]

// DO NOT MODIFY THIS LINE. </GHA-BLOG-TOPIC> 

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // // 
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/22950
// Contact/Reviewer: ReeceNana,cgianfra
// // // // // // // // 
[#batch]
== Improved CDI integration (Jakarta Batch 2.1)
The updates in Jakarta Batch 2.1 primarily focus on improved integration with Context Dependency Injection (CDI). Discovery of Batch artifacts is now standardized via CDI, whereas before it was not fully defined in the specification.

Previously, the default `JobOperator` was only made available from a static factory method, `BatchRuntime.getJobOperator()`. With Jakarta Batch 2.1, the default `JobOperator` is provided as a CDI bean unless one is defined by the user application, if an injectable field is available as defined by: 
[source,java]
----
`@Inject` 

> `JobOperator jobOperator;`

Reviewer 4: ```suggestion
@Inject 
JobOperator jobOperator;

```

----

Finally, the `@BatchProperty` annotation allowed for injection of String-type values only, but has been expanded to most other primitive types. For example:
[source,java]
----

> `@Inject @BatchProperty(name="p1") String p1;`

Reviewer 4: ```suggestion
@Inject @BatchProperty(name="p1") String p1;

```

----

This property can have other types in Jakarta Batch 2.1, including `Boolean`, `Integer`, `Long`, `Float`, and `Double`. For example:
[source,java]
----

> `@Inject @BatchProperty(name="p1") Integer p1;`

Reviewer 4: ```suggestion
@Inject @BatchProperty(name="p1") Integer p1;

```

To enable the new Jakarta Messaging 4.0 feature, add the `messaging-3.1` feature to your server.xml:
[source,xml]
---- 
<featureManager>
   <feature>messaging-3.1</feature>
</featureManager>   
----

For more information about Jakarta Messaging, refer to the following links:  

* link:{url-prefix}/docs/latest/reference/feature/messaging.html[Feature Documentation: Jakarta Messaging]
* link:https://jakarta.ee/specifications/messaging/3.1/jakarta-messaging-spec-3.1.html[Specification: Jakarta XML Web Services]

// DO NOT MODIFY THIS LINE. </GHA-BLOG-TOPIC> 

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // // 
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/22902
// Contact/Reviewer: neuwerk,ReeceNana
// // // // // // // // 
[#xmlws]
== Jakarta Web Services Metadata incorporated into XML Web Services 4.0

> Jakarta XML Web Services 4.0 is a set of APIs that allows applications to deploy SOAP Based Web Services on the Jakarta EE 10 platform. This set of APIs includes both the XML Web Services 4.0 specification which now incorporates the Jakarta Web Services Metadata specification, as well as the Jakarta SOAP with Attachments 3.0 specification. 

Reviewer: ```suggestion
Jakarta XML Web Services 4.0 is a set of APIs that allows applications to deploy SOAP-based web services on the Jakarta EE 10 platform. This set of APIs includes both the XML Web Services 4.0 specification, which now incorporates the Jakarta Web Services Metadata specification, as well as the Jakarta SOAP with Attachments 3.0 specification. 

```

> Although the changes introduced in XML Web Services 4.0 are relatively minor, they do include the requirement for Java SE 11 as well as incorporating the Jakarta Web Services Metadata specification into XML Web Services 4.0. 

Reviewer: ```suggestion
Although the changes introduced in XML Web Services 4.0 are relatively minor, they do include the requirement for Java SE 11, as well as incorporating the Jakarta Web Services Metadata specification into XML Web Services 4.0. 

```

To enable the new Jakarta XML Web Services 4.0 feature, add the `xmlWS-4.0` feature to your server.xml:
[source,xml]
---- 
<featureManager>
   <feature>xmlWS-4.0</feature>
</featureManager>   
----

For more information about Jakarta XML Web Services, refer to the following links:  

* link:{url-prefix}/docs/latest/reference/feature/xmlWS.html[Feature Documentation: Jakarta XML Web Services]
* link:https://jakarta.ee/specifications/xml-web-services/4.0/jakarta-xml-ws-spec-4.0.html[Specification: Jakarta XML Web Services]

// DO NOT MODIFY THIS LINE. </GHA-BLOG-TOPIC> 

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // // 
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/22930
// Contact/Reviewer: teddyjtorres,ReeceNana
// // // // // // // // 
[#jwt]
== Simplified JWT Validation (MicroProfile JSON Web Token 2.1)

> The MicroProfile JSON Web Token 2.1 specification allows the use of a JWT token for authenticating and authorizing requests to a service. The specification simplifies the configuration for managing the validation of the JWT by introducing three new MicroProfile Config properties along with corresponding `mpJwt` attributes.

Reviewer: ```suggestion
The MicroProfile JSON Web Token 2.1 specification allows the use of a JWT to authenticate and authorize requests to a service. The specification simplifies managing the validation of the JWT by introducing three new MicroProfile Config properties, along with corresponding `mpJwt` attributes.

```

=== mp.jwt.verify.token.age

> The `mp.jwt.verify.token.age` property allows you to specify the number of **seconds** since the JWT token was issued. The "iat" claim must be present in the JWT and the configured number of seconds since "iat" must not have elapsed. If it has elapsed, then the request is rejected with an Unauthorized (401) response.

Reviewer: ```suggestion
The `mp.jwt.verify.token.age` property allows you to specify the number of **seconds** since the JWT token was issued. The `iat` (issued at) claim must be present in the JWT and the configured number of seconds since `iat` must not have elapsed. If it has elapsed, then the request is rejected with an Unauthorized (401) response.

```

  
You can enable the MicroProfile JSON Web Token 2.1 feature by using the `mpJwt-2.1` feature in the server.xml file: 
[source,xml]
----
<featureManager>
   <feature>mpJwt-2.1</feature>
</featureManager>
----

For more information regarding MicroProfile JSON Web Token, refer to the following links:  

* link:{url-prefix}/guides/microprofile-jwt.html[Guide: Securing microservices with JSON Web Tokens]
* link:{url-prefix}/docs/latest/reference/feature/mpJwt.html[Feature Documentation: MicroProfile JSON Web Token]
* link:{url-prefix}/docs/latest/microprofile-config-properties.html#jwt[Configuration Documentation: MicroProfile JSON Web Token]
* link:https://github.com/eclipse/microprofile-jwt-auth/blob/master/spec/src/main/asciidoc/microprofile-jwt-auth-spec.asciidoc[Specification: MicroProfile JWT]

// DO NOT MODIFY THIS LINE. </GHA-BLOG-TOPIC> 

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // // 
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/22917
// Contact/Reviewer: ReeceNana,Channyboy
// // // // // // // // 
[#metrics]

> == User defined metric registry scopes (MicroProfile Metrics 5.0)

Reviewer: ```suggestion
== User-defined metric registry scopes (MicroProfile Metrics 5.0)

```

> The MicroProfile Metrics 5.0 feature is the implementation of the Eclipse MicroProfile Metrics 5.0 release and is only be compatible to run with other Jakarta EE 10 features. This latest version of MicroProfile Metrics brings with it both new functionality as well as significant changes to the functionality in MicroProfile Metrics 4.0 (i.e. the Open Liberty `mpMetrics-4.0` feature).  

Reviewer: ```suggestion
The MicroProfile Metrics 5.0 feature is the implementation of the Eclipse MicroProfile Metrics 5.0 release and is compatible to run only with other Jakarta EE 10 features. This latest version of MicroProfile Metrics includes both new functionality and significant changes to the functionality in MicroProfile Metrics 4.0 (i.e. the Open Liberty `mpMetrics-4.0` feature).  

```

> Before we cover the new, lets first quickly go over the changes to existing functionality.  Most notable is the removal of the `Meter`, `ConcurrentGauge` and `SimpleTimer` metrics. A consequence of this change is that the `REST.request` metric, which was previously backed by a `SimpleTimer` metric, is now a `Timer` metric. For the publishing of metrics, only the Prometheus format is now available, and the JSON formatted output has been removed. Additionally, the metrics endpoint for specific metric registry scope and metric is no longer accessed through a tree hierarchy, but rather through query params. For example, `/metrics/base/jvm.uptime` will be accessed through  `/metrics?scope=base&metric=jvm.uptime`. The `/metrics?scope=base` endpoint by itself will retrieve the metrics for the base metric registry.

Reviewer: ```suggestion
Before we cover the new, lets first quickly go over the changes to existing functionality.  Most notable is the removal of the `Meter`, `ConcurrentGauge` and `SimpleTimer` metrics. A consequence of this change is that the `REST.request` metric, which was previously backed by a `SimpleTimer` metric, is now a `Timer` metric. For the publishing of metrics, only the Prometheus format is now available, and the JSON formatted output has been removed. Additionally, the metrics endpoint for a specific metric registry scope and metric is no longer accessed through a tree hierarchy, but rather through query params. For example, `/metrics/base/jvm.uptime` is accessed through  `/metrics?scope=base&metric=jvm.uptime`. The `/metrics?scope=base` endpoint by itself retrieves the metrics for the base metric registry.

```

> New in this release is the introduction of user-defined custom scopes. Prior to this release, there was only the base, vendor and application scopes and only the application scope could be used to register metrics. Now, metrics can be registered to a user defined metric registry scope in one of two ways:

Reviewer: ```suggestion
New in this release is the introduction of user-defined custom scopes. Prior to this release, there was only the base, vendor and application scopes and only the application scope could be used to register metrics. Now, metrics can be registered to a user-defined metric registry scope in one of the following two ways.

```

> ==== 1. Scope attribute in a the metric annotations

Reviewer: ```suggestion
==== Scope attribute in a the metric annotations

```

[source,xml]
----
@Counted(name = "myCounter", scope ="customScope")
public void foo(){
   ...
}
----

> ==== 2. Inject a metric registry with a custom scope using the new `@RegistryScope` annotation

Reviewer: ```suggestion
==== Inject a metric registry with a custom scope using the new `@RegistryScope` annotation

```

[source,xml]
----
@Inject 
@RegistryScope(scope="customScope")
MetricRegistry metricRegistry;
----
NOTE: The `@RegistryType` injection qualifier is still available, but is now deprecated.

> Furthermore, the `mp.metrics.smallrye.timer.precision` and `mp.metrics.smallrye.histogram.precision` MicroProfile Config values have been introduced in this release. These values specify the percentile accuracy used by the `Timer` and `Histogram` metrics.  The valid values are 1-5. By default, the configuration is `3`. The higher the number, the greater the precision, but the greater the memory usage.

Reviewer: ```suggestion
Furthermore, the `mp.metrics.smallrye.timer.precision` and `mp.metrics.smallrye.histogram.precision` MicroProfile Config properties are introduced in this release. These properties specify the percentile accuracy used by the `Timer` and `Histogram` metrics.  The valid values are 1-5. The default value is `3`. The higher the number, the greater the precision, but the greater the memory usage.

```

> The above are only the most notable and impactful changes. There are further changes to the API which have not been covered in this blog.  For a comprehensive list of changes to the MicroProfile Metrics technology, go to the project page at [Eclipse MicroProfile Metrics project](https://github.com/eclipse/microprofile-metrics)

Reviewer: ```suggestion
These only the most notable and impactful changes. There are further changes to the API that are not covered in this blog.  For a comprehensive list of changes to the MicroProfile Metrics technology, go to the link:https://github.com/eclipse/microprofile-metrics[Eclipse MicroProfile Metrics project page].

```

blog_description: "Open Liberty continues to make progress towards full Jakarta EE 10 compliance with the debut of six new specification implementations Jakarta WebSocket 2.1, Jakarta Batch 2.1, Jakarta XML Web Services 4.0, Jakarta Messaging 3.1, Jakarta Messaging 3.1, Jakarta Pages 3.1 and Jakarta Tags 3.0.  Two new MicroProfile 6 features are also introduced, MicroProfile JSON Web Token 2.1 and MicroProfile Metrics 5.0."
open-graph-image: https://openliberty.io/img/twitter_card.jpg
open-graph-image-alt: Open Liberty Logo
---
= More Jakarta EE 10 and MicroProfile 6 features make their debut in Open Liberty 22.0.0.12-beta!
Michal Broz <https://github.com/mbroz2>
:imagesdir: /
:url-prefix:
:url-about: /
//Blank line here is necessary before starting the body of the post.

> As we continue to make progress towards full Jakarta EE 10 compliance, the 22.0.0.12-beta release brings with it six new Jakarta EE 10 specifications, alongside the ones which have been part of previous beta releases.  Jakarta Batch 2.1 provides improved CDI integration while Jakarta XML Web Services 4.0 now incorporates Jakarta Web Services Metadata.  Jakarta Pages 3.1 provides developers an option to raise Exceptions for undefined Expression Language Identifiers, while Jakarta Tags 3.0 includes Tag URI renaming when referencing tag libraries. Jakarta Messaging 3.1 makes connection factory and destination definition annotations repeatable, while Jakarta WebSocket 2.1 allows for the registration of endpoints outside of the deployment phase.

Reviewer 2: ```suggestion
As we progress towards full Jakarta EE 10 compliance, the 22.0.0.12-beta release includes six new Jakarta EE 10 specifications, alongside the ones that were included in previous beta releases.  

Jakarta Batch 2.1 provides improved CDI integration while Jakarta XML Web Services 4.0 now incorporates Jakarta Web Services Metadata.  Jakarta Pages 3.1 provides developers an option to raise Exceptions for undefined Expression Language Identifiers, while Jakarta Tags 3.0 includes Tag URI renaming when referencing tag libraries. Jakarta Messaging 3.1 makes connection factory and destination definition annotations repeatable, while Jakarta WebSocket 2.1 allows for the registration of endpoints outside of the deployment phase.

```
